### DISCUSSION TITLE
Advice regarding allowSomeUserActionsEvenWhenInCardRoutine

### DISCUSSION THREAD

**Discussion by RJ-Eckie** [SynthstromAudible/DelugeFirmware#2940] [General]
Does anyone know the purpose of the mysterious `allowSomeUserActionsEvenWhenInCardRoutine `? In *audio_file_manager.cpp:*

<img width="831" alt="Screenshot 2024-11-19 at 21 03 03" src="https://github.com/user-attachments/assets/54f11385-e9c9-4f9c-812a-d488cfead325">

By default UI input is ignored while in `sdRoutineLock` but there is a focused effort in the code to have the vertical encoder still be interpreted during `sdRoutineLock` if `allowSomeUserActionsEvenWhenInCardRoutine` is true.

This makes it currently impossible to simplify the UI code by handling `sdRoutineLock` in a central place. I'd like to understand what exactly is happening here so I can create a clean implementation for this exception - or remove it if it's no longer used. Thanks 🙏🏻

**Comment by stellar-aria** [comment]
It allows processing of some UI rendering even while inside the card routine, since the card routine also does input event handling. See the views, such as https://github.com/SynthstromAudible/DelugeFirmware/blob/be28b94a08112ead22cb6e10ae3b122e23d6be9b/src/deluge/gui/views/audio_clip_view.cpp#L781-L795 for an example.

The SD card routine is run on such a regular basis (and is apparently long enough cycle-wise) that not handling user input events during it would potentially impact responsiveness.

Solutions for this would be splitting up the larger monolithic sdCardRoutine into smaller tasks that can have input event handling done in between them, and maybe implementing a better event queue system. I'm not sure how we currently handle user events (whether queued or immediate).

**Reply by RJ-Eckie** [reply]
I kind of figured that would be its most likely purpose. But if this is the case, why is it only implemented on the vertical encoder and not on any of the other controls? All non-vertical encoders don't even have a `bool inCardRoutine` parameter and are always skipped during `sdRoutineLock`:
https://github.com/SynthstromAudible/DelugeFirmware/blob/be28b94a08112ead22cb6e10ae3b122e23d6be9b/src/deluge/gui/ui/ui.h#L100-L103
There are only two exceptions to this: [pressing an audition pad in clip view](https://github.com/SynthstromAudible/DelugeFirmware/blob/02dba129bd4a7c9a7fef40c8425590b90e22a436/src/deluge/gui/views/instrument_clip_view.cpp#L4477-L4481) and [pressing any pad in keyboard screen](https://github.com/SynthstromAudible/DelugeFirmware/blob/02dba129bd4a7c9a7fef40c8425590b90e22a436/src/deluge/gui/ui/keyboard/keyboard_screen.cpp#L94-L97).

Outside of those exceptions, none of the grid pads ever respond to `allowSomeUserActionsEvenWhenInCardRoutine`. Neither do five of the six encoders. From a responsiveness point of view, would this even have any impact if such a large part of the UI isn't taken into consideration? Could this be the start of a responsiveness-fix that was never implemented beyond the first stage?

**Reply by RJ-Eckie** [reply]
> I'm not sure how we currently handle user events (whether queued or immediate).

User events are handled by the task manager with different priorities:
https://github.com/SynthstromAudible/DelugeFirmware/blob/02dba129bd4a7c9a7fef40c8425590b90e22a436/src/deluge/deluge.cpp#L578
https://github.com/SynthstromAudible/DelugeFirmware/blob/02dba129bd4a7c9a7fef40c8425590b90e22a436/src/deluge/deluge.cpp#L590
I have focused on encoders so far: they are read very often to avoid skipping steps. Interpreting the resulting values and sending them to the appropriate UI/View happens with lower frequency. In the vast majority of cases, the second step -`encoders::interpretEncoders()` - is simply ignored when `sdRoutineLock` is true.

**Reply by stellar-aria** [reply]
Yeah the situation with the encoders isn't great because of how they're connected to the CPU. I don't know if allowSomeUserActionsEvenWhenInCardRoutine is a remnant of workarounds or not.

We've had some discussion, and if you'd like to switch the encoder system to being driven by pin-change interrupts so we can get rid of the polling complications, that'd probably be the best move forward.

Keep track of very minimal pin state per encoder, and every time there's a full quadrature cycle enqueue some kind of encoder event to be processed later.

```c++
struct Encoder {
  void processA(bool quadratureA) {
    quadrature_a_ = (quadrature_a_ << 1) | static_cast<int>(quadratureA);
    quadrature_b_ = (quadrature_b_ << 1) | (quadrature_b_ & 0b1);
    handle();
  }
  void processB(bool quadratureB) {
    quadrature_b_ = (quadrature_b_ << 1) | static_cast<int>(quadratureB);
    quadrature_a_ = (quadrature_a_ << 1) | (quadtrature_a_ & 0b1);
    handle();
  }
  void handle() const {
    uint8_t a = quadrature_a_ & 0b11;  // only care about last two samples
    uint8_t b = quadrature_b_ & 0b11;
    if (a == 0b10 && b == 0b00) { // was 1,0 now 0,0
      Event::enqueue(name, 1);
    }
    else if (a == 0b00 && b == 0b10) { // was 0,1 now 00
      Event::enqueue(name, -1);
    } else {
        // still moving to detent, do nothing
    }
  }
  EncoderName name;
  uint8_t quadrature_a_ = 0x00;
  uint8_t quadrature_b_ = 0x00;
};
``` 
kinda thing

note: edited to fix bug where history doesn't update for other pin when one pin is processed

**Reply by RJ-Eckie** [reply]
Where the code is triggered from isn't really my issue. I've currently rewritten the entire encoder-side of the ui and I have no issue calling it from the exact same position it's currently called from. (Replacing the community-tasks with mine.) Calling it from an interrupt sounds like a good idea - but that is separate from the issue I'm running into.

My problem is that I want to create a simple and central check against `sdRoutineLock`. Currently I'd say 98-99% of user input is not processed when `sdRoutineLock` is true. I would like to make that 100% by creating a central check and just not calling any of the interpretEncoder()/padAction()/buttonAction() (etc.) code at all during `sdRoutineLock`. We would need a total of two checks: one for handling presses (pads/buttons/encoders) and one for handling encoder rotations. This simplification is currently not possible because of the `allowSomeUserActionsEvenWhenInCardRoutine` exceptions. Currently there are almost a hundred checks for `sdRoutineLock` in the codebase. 

So my concrete question is: seeing that `allowSomeUserActionsEvenWhenInCardRoutine` is used in only a few places, with no apparent consistency, and I would say in the current implementation with no real effect on responsiveness - can we just delete it? Removing these very specific and rare exceptions would allow for `sdRoutineLock` to just be a 100% lock for processing user input.

**Reply by stellar-aria** [reply]
I would say try it, and if it doesn't have any horrendous obvious bugs we'll likely upstream it and try to handle any resulting issues architecturally, yes.

**Reply by seangoodvibes** [reply]
> Where the code is triggered from isn't really my issue. I've currently rewritten the entire encoder-side of the ui and I have no issue calling it from the exact same position it's currently called from. (Replacing the community-tasks with mine.) Calling it from an interrupt sounds like a good idea - but that is separate from the issue I'm running into.
> 
> 
> 
> My problem is that I want to create a simple and central check against `sdRoutineLock`. Currently I'd say 98-99% of user input is not processed when `sdRoutineLock` is true. I would like to make that 100% by creating a central check and just not calling any of the interpretEncoder()/padAction()/buttonAction() (etc.) code at all during `sdRoutineLock`. We would need a total of two checks: one for handling presses (pads/buttons/encoders) and one for handling encoder rotations. This simplification is currently not possible because of the `allowSomeUserActionsEvenWhenInCardRoutine` exceptions. Currently there are almost a hundred checks for `sdRoutineLock` in the codebase. 
> 
> 
> 
> So my concrete question is: seeing that `allowSomeUserActionsEvenWhenInCardRoutine` is used in only a few places, with no apparent consistency, and I would say in the current implementation with no real effect on responsiveness - can we just delete it? Removing these very specific and rare exceptions would allow for `sdRoutineLock` to just be a 100% lock for processing user input.
> 
> 

You can certainly try this.

But I'm not sure if centralizing that check and not calling the action functions while in the routine is the best solution.

I did that in automation view (by putting the check at the top of the button and pad action functions) because I thought that would simplify the logic

But upon benchmarking it, it was discovered that automation view's button and pad action functions were a lot slower than other button and pad actions while in the card routine.

The recommendation was to allow fast processor actions to happen while in the card routine and delay other slower actions from happening while in the card routine.

I believe Rohan has done some sort of exercise to determine what actions are fast and slow but this has not been reviewed since. This is evidenced by the in card routine check not being checked everywhere in the action functions.

Ultimately the objective here is to determine what actions when combined with being in the card routine would not slow down the deluge's performance as the deluge's performance is a combination of how long it takes to render and sd card streaming.

"You’ll get culling when rendering 128 samples is taking longer than outputting them or when the sd card can’t read the next cluster before it’s needed"

**Reply by RJ-Eckie** [reply]
> The recommendation was to allow fast processor actions to happen while in the card routine and delay other slower actions from happening while in the card routine.

I can totally see that, but what you're describing is different from the results of `allowSomeUserActionsEvenWhenInCardRoutine`. (I'm just gonna call it `allowSome` for now.)

What you're describing is fully depends on `sdRoutineLock`:
- some code should be blocked during card routine
- some fast code should run during card routine

(Totally fair, and I'd probably personally create a centralised way of handling that as well - but for now I just want to focus on `allowSome` for a second.)

`allowSome` defines a different behavior:
- some code should always be blocked during card routine
https://github.com/SynthstromAudible/DelugeFirmware/blob/92a67bf21e400b4a38491594fd9144eb189aa4fc/src/deluge/gui/context_menu/midi_learn_mode.cpp#L85-L87
- some code should be blocked during card routine _except_ during `AudioFileManager::loadAnyEnqueuedClusters() { loadCluster(cluster); }`
https://github.com/SynthstromAudible/DelugeFirmware/blob/92a67bf21e400b4a38491594fd9144eb189aa4fc/src/deluge/gui/ui/keyboard/keyboard_screen.cpp#L650-L652

I'm just trying to work out that specific and rare behavior that `allowSome` enforces and my argument so far is that it just doesn't really do anything noticable. I'm keeping my branch 100% compatible with `community` so if we could agree on a consistent implementation of `allowSome`, _or_ that it does so little that it can just be removed, that would make my life considerably easier.

For now I guess I'm just setting up a switchable test-case to turn it off and see whether anything breaks!

**Reply by nikodemus** [reply]
I was at some point considering changing those `if`s to:
```
#define MAYBE_BLOCK(X) if (shouldBlock(X)) { return REMIND_ME_LATER; }
```
where X would be some kind of action-priority/heaviness enum. Would be also interesting to test the effect of adding static branch predictions either way to those.

**Reply by RJ-Eckie** [reply]
> I was at some point considering changing those `if`s to:
> 
> ```
> #define MAYBE_BLOCK(X) if (shouldBlock(X)) { return REMIND_ME_LATER; }
> ```
> 
> where X would be some kind of action-priority/heaviness enum. Would be also interesting to test the effect of adding static branch predictions either way to those.

I think it's useful to separate the two different discussions that are going on here: 

1. How to deal with priority and/or code blocking during the SD routine is a very interesting and possibly complex discussion. The current setup to my understanding is as follows:
- The task manager manages proirity issues, dealing with hardware routines vs more high-level ui interactions
- During the SD routine, a bunch of code just doesn't get executed to prevent clashes. This isn't always implemented in the most consistent way, but it looks like that is the general design.

   I think implementing that system in a consistent way and evaluating the results of that would at this point be a more efficient way of spending effort than creating a new, more complex system - which we aren't sure we need without the evaluation of the current system. That's and interesting conversation to have

**However** I would like to separate that complex conversation from the way more simple question I'm asking here. Which is:

2. Does the current implementation of `allowSomeUserActionsEvenWhenInCardRoutine` do anything, at all? I'm arguing it doesn't.

   Agreeing on that answer would allow us to remove a range of checks that create a significant amount of chaos in UI code as it is currently, without making any changes to the priority and blocking behavior. The conversation about fundamental changes to the priority and blocking implementation would be a separate one
